modules/placement-driver/src/main/java/org/apache/ignite/internal/raft/client/TopologyAwareRaftGroupService.java

```diff
@@ -0,0 +1,443 @@
+     *
+     * @param node Node.
+     * @param msg  Subscribe message.
+     * @return A future that will complete when the message is sent.
+     */
+    private CompletableFuture<Boolean> sendSubscribeMessage(ClusterNode node, SubscriptionLeaderChangeRequest msg) {
+        var msgSendFut = new CompletableFuture<Boolean>();
+
+        sendWithRetry(node, msg, msgSendFut);
+
+        return msgSendFut;
+    }
+
+    /**
+     * Tries to send a subscribe message until the node leaves of the cluster.
+     *
+     * @param node       Node.
+     * @param msg        Subscribe message to send.
+     * @param msgSendFut Future that will completed when the message is send or an issue prevent that.
+     */
+    private void sendWithRetry(ClusterNode node, SubscriptionLeaderChangeRequest msg, CompletableFuture<Boolean> msgSendFut) {
+        clusterService.messagingService().invoke(node, msg, raftConfiguration.responseTimeout().value()).whenCompleteAsync((unused, th) -> {
+            if (th != null) {
+                if (recoverable(th)) {
+                    logicalTopologyService.logicalTopologyOnLeader().whenCompleteAsync((logicalTopologySnapshot, topologyGetIssue) -> {
+                        if (topologyGetIssue != null) {
+                            LOG.error("Actual logical topology snapshot was not get.", topologyGetIssue);
+
+                            msgSendFut.completeExceptionally(topologyGetIssue);
+
+                            return;
+                        }
+
+                        if (logicalTopologySnapshot.nodes().contains(node)) {
+                            sendWithRetry(node, msg, msgSendFut);
+                        } else {
+                            LOG.info("Could not subscribe to leader update from a specific node, because the node had left of the cluster"
+                                    + " [node={}]", node);
+
+                            msgSendFut.complete(false);
+                        }
+                    }, executor);
+                } else {
+                    LOG.error("Could not send the subscribe message to the node [node={}, msg={}]", th, node, msg);
+
+                    msgSendFut.completeExceptionally(th);
+                }
+
+                return;
+            }
+
+            msgSendFut.complete(true);
+        }, executor);
+    }
+
+    /**
+     * Whether an exception is recoverable or not.
+     *
+     * @param t The exception.
+     * @return True if the exception is recoverable, false otherwise.
+     */
+    private static boolean recoverable(Throwable t) {
+        if (t instanceof ExecutionException || t instanceof CompletionException) {
+            t = t.getCause();
+        }
+
+        return t instanceof TimeoutException || t instanceof IOException;
+    }
+
+    /**
+     * Assigns a closure to call when a leader will is elected.
+     *
+     * @param callback Callback closure.
+     */
+    public void subscribeLeader(Consumer<ClusterNode> callback) {
```

Comment:
Every call replaces previous callback, is it correct?

Reply:
Do you mean need to check that the service have not already subscribed before the method is invoked?

Reply:
Added assert here.
assert !serverEventHandler.isSubscribed() : "The node already subscribed";